### PR TITLE
Softmax unit-test reduction across all and non innermost dims cases.

### PR DESCRIPTION
**Pull request by aosewski**
This PR adds cases checking reduction across all axis. They're commented-out for now, since they're failing. So this is a reminder to fix this case in future.

- [x] Test ckProfiler
- [x] Have instances for corner cases in separate cpp file.

### CODE REVIEW

**Comment by aosewski** [comment]
@rosenrodt Just FYI: right now also reduction on other than innermost dimension also fails. But I couldn't set-up such configuration in current unit-test implementation.

**Comment by asroy** [comment]
@aosewski I think internally, all N-D softmax is converted into 2-D softmax, by merging dimensions. In that case, it's easier just to assume the user input tensor will be at least 2D. If user wants to reduce all dimensions of a ```N-D``` tensor, they can just make a N+1 dimensional tensor

In that case, do we still need these tests?

**Comment by rosenrodt** [comment]
The failing cases are due to oversight in host softmax implementation. Tentative fix. @aosewski could you try cherry-picking https://github.com/rosenrodt/composable_kernel/commit/71e9d7345e16f1398c8f7897f06b75a3f1123051 and enable the reduce all dims test?

**Comment by aosewski** [comment]
> @aosewski I think internally, all N-D softmax is converted into 2-D softmax, by merging dimensions. In that case, it's easier just to assume the user input tensor will be at least 2D. If user wants to reduce all dimensions of a `N-D` tensor, they can just make a N+1 dimensional tensor
> 
> In that case, do we still need these tests?

Well, actually turns out that those test cases pass, so why not leaving them?

**Comment by aosewski** [comment]
> The failing cases are due to oversight in host softmax implementation. @aosewski could you try cherry-picking the fix [rosenrodt@71e9d73](https://github.com/rosenrodt/composable_kernel/commit/71e9d7345e16f1398c8f7897f06b75a3f1123051) and enable the reduce all dims test?

@rosenrodt I've enabled those tests. Moreover I've added test cases with reduction on non innermost dimension - although they're not failing, they're all actually silently skipped because of not supported argument.

File: test/softmax/test_softmax_fp16.cpp
Comment (rosenrodt) [inline test/softmax/test_softmax_fp16.cpp:61]: I think TestSoftmaxFP16_v2 can be merged into TestSoftmaxFP16, then we can skip the test when requested `reduce_dims.size()` doesn't match `NumReduceDim` in the kernel

**Review by rosenrodt — APPROVED** [review]
CI failure is blocked by entirely different issue since #399 was merged. It should be fixed in #409 which is still under CI test.

File: test/softmax/test_softmax_fp16.cpp
Comment (rosenrodt) [reply test/softmax/test_softmax_fp16.cpp:61]: Or even if there's a mismatch, maybe the `IsSupportedArgument` should catch the mismatch and simply skip it. I think I didn't properly implement the check before

https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/ce74cea4073067f6784ebbc73f4d565662c98125/include/ck/tensor_operation/gpu/device/impl/device_softmax_impl.hpp#L194-L209

**Comment by aosewski** [comment]
@rosenrodt I merged both test fixtures into sth like this:
```
TYPED_TEST(TestSoftmaxFP16, ReduceInnermostDim)
{
    std::vector<ck::index_t> reduce_dims{2};
    this->Run(reduce_dims);
}
TYPED_TEST(TestSoftmaxFP16, ReduceInnerDim)
{
    std::vector<ck::index_t> reduce_dims{1};
    this->Run(reduce_dims);
}
TYPED_TEST(TestSoftmaxFP16, ReduceOutermostDim)
{
    std::vector<ck::index_t> reduce_dims{0};
    this->Run(reduce_dims);
}
TYPED_TEST(TestSoftmaxFP16, Reduce12Dims)
{
    std::vector<ck::index_t> reduce_dims{1, 2};
    this->Run(reduce_dims);
}
TYPED_TEST(TestSoftmaxFP16, Reduce01Dims)
{
    std::vector<ck::index_t> reduce_dims{0, 1};
    this->Run(reduce_dims);
}
TYPED_TEST(TestSoftmaxFP16, Reduce02Dims)
{
    std::vector<ck::index_t> reduce_dims{0, 2};
    this->Run(reduce_dims);
}
TYPED_TEST(TestSoftmaxFP16, ReduceAllDims)
{
    std::vector<ck::index_t> reduce_dims{0, 1, 2};
    this->Run(reduce_dims);
}
```
But this resulted in plethora of test failures. What turned out, there's a bug in this line: github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/include/ck/tensor_operation/gpu/device/device_reduce_common.hpp#L80 Which lets the user to pass the vector of reduction dims of different size than template parameter `NumReduceDim`. That's because in `Release` build there is `NDEBUG` macro defined which changes `assert` macro into no op, thus this check is not tested at all. This condition is check at this place at run-time, thus we can't use `static_assert` - the best way IMHO is to check it before in `DeviceReduceMultiBlock::Argument` constructor and throw an error if not satisfied. 

@asroy  I want to put your attention to similar `assert`s statements as above, since quick search showed that we may have few of them which may check important conditions, but due to `NDEBUG` macro are not checked at all in `Release` builds.

**Comment by aosewski** [comment]
@rosenrodt I've reorganized test cases into three test suites wrt. `NumReduceDim` - this allow to not duplicate test cases, and allow to check for expected failures: ie. when `NumReduceDim=2` and user passed `reduce_dims{1}` that is when the number of dimensions to reduce doesn't equal to template parameter `NumReduceDim`. We should actually also check that the requested dimensions to reduce are within `(0, Rank]` range.

**Comment by rosenrodt** [comment]
@aosewski Awesome, thanks!

> But this resulted in plethora of test failures. What turned out, there's a bug in this line:
> 
> https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/include/ck/tensor_operation/gpu/device/device_reduce_common.hpp#L80
> 
> Which lets the user to pass the vector of reduction dims of different size than template parameter `NumReduceDim`. That's because in `Release` build there is `NDEBUG` macro defined which changes `assert` macro into no op, thus this check is not tested at all. The condition is checked only at run-time, thus we can't use `static_assert.` The best way IMHO is to check it before in `DeviceReduceMultiBlock::Argument` constructor and throw an error if not satisfied.
> 
> @asroy I want to put your attention to similar `assert`s statements as above, since quick search showed that we may have few of them which may check important conditions, but due to `NDEBUG` macro are not checked at all in `Release` builds.

In fact, in current develop branch there's a lot of `assert()`'s triggered resulting in test failures when NDEBUG is not defined. So I second that.

**Comment by rosenrodt** [comment]
@aosewski Please fix error reported by CI

```
[2022-09-12T13:04:32.680Z] [ 93%] Building CXX object test/softmax/CMakeFiles/test_softmax_fp32.dir/test_softmax_fp32.cpp.o
[2022-09-12T13:04:35.215Z] /var/jenkins/workspace/sable_kernel_aosewski_softmax_ut/test/softmax/test_softmax_fp32.cpp:71:2: error: no newline at end of file [-Werror,-Wnewline-eof]
[2022-09-12T13:04:35.215Z] }
[2022-09-12T13:04:35.215Z]  ^
[2022-09-12T13:04:43.335Z] 1 error generated when compiling for gfx1030.
[2022-09-12T13:04:43.335Z] test/softmax/CMakeFiles/test_softmax_fp32.dir/build.make:62: recipe for target 'test/softmax/CMakeFiles/test_softmax_fp32.dir/test_softmax_fp32.cpp.o' failed
```

**Comment by aosewski** [comment]
@rosenrodt Fixed. And reorganized unit-tests to check also reduction across inner dimensions - this required custom op instances.

File: test/softmax/test_softmax_fp32.cpp
Comment (rosenrodt) [inline test/softmax/test_softmax_fp32.cpp:105]: Middle/outer dimensions might be a bit more accurate?

**Review by rosenrodt — APPROVED** [review]
I like these refined unit tests. It is very thorough.

**Comment by rosenrodt** [comment]
@aosewski Is it better if softmax UTs follow current practice where test programs call kernels from the static library containing all the kernel instances?

**Comment by aosewski** [comment]
> @aosewski Is it better if softmax UTs follow current practice where test programs call kernels from the static library containing all the kernel instances?

That would make sense if we would have defined set of instances which covers same functionality. Otherwise a lot of code-paths won't be covered.

**Comment by aosewski** [comment]
@asroy @rosenrodt The PR is ready to review.

File: test/softmax/test_softmax_interface.cpp
Comment (rosenrodt) [inline test/softmax/test_softmax_interface.cpp:10]: Separating interface testing from verification under different data types is a good one.

File: test/softmax/test_softmax_interface.cpp
Comment (rosenrodt) [inline test/softmax/test_softmax_interface.cpp:23]: Where does DeviceSoftmaxImpl throw? Shouldn't we be expecting a boolean here? 

Assuming it does throw from somewhere, I suggest replacing EXPECT_THROW with EXPECT_FALSE for all tests checking against IsSupportedArgument() and avoid throwing exception in IsSupportedArgument(). The softmax profiler code (as well as most other ops' profiler codes in CK) does not expect to handle exceptions. 

There is a valid case when it is more desirable to silently skip unsupported kernels. Such as the case that we want to benchmark a specific GEMM size against a pool of kernel instances that may or may not support padding. Or likewise in softmax that an arbitrary input tensor that doesn't satisfy the condition of aligned vectorized load/store (SrcScalarPerVector) shouldn't cause the profiler to come to a full stop.

**Comment by rosenrodt** [comment]
@aosewski Huge thanks for improving the usability of standalone softmax kernel. Especially the robust interface tests you recently added.

File: test/softmax/test_softmax_interface.cpp
Comment (aosewski) [reply test/softmax/test_softmax_interface.cpp:23]: @rosenrodt I understand your concerns regarding throwing an exception. However I couldn't find any better solution to reject creating an `Argument` for DeviceOp when the user pass ie. reduction axis which are out of input tensor bounds. Take a look here: https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/4b4b66756cecc5b57b4453f9a5015c4a603d54b3/include/ck/tensor_operation/gpu/device/device_reduce_multiblock.hpp#L223-L245

Previously without those checks the user could run such illformed case since in this place it isn't checked correctly: https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/include/ck/tensor_operation/gpu/device/device_reduce_common.hpp#L75-L105 - this is the incorrect use of `assert` which is in Release build mode a no op.

Comment (rosenrodt) [reply test/softmax/test_softmax_interface.cpp:23]: Thanks for explanation. I see what's going on here—It's not IsSupportedArgument() who throws, but MakeArgument().

Two scenarios where we can't run a problem size against specific kernel type:
1. **Unsupported**: problem may match at least one kernel, because some kernels have their own implementation details which prevent all kernels from running correctly (eg GEMM MNK padding, ScalarPerVector)
2. **Incorrect**: problem does not match any kernel at all (reducing 2 dimensions when kernel only supports 1 dimension reduction)

The cases that fall into incorrect scenario should throw an exception as it's obviously user error. The cases that fall into unsupported scenario is not user error. All look good to me now.

**Review by rosenrodt — DISMISSED** [review]
LGTM

**Comment by rosenrodt** [comment]
Looping in @qianfengz as he is currently modifying some part of softmax & multiblock reduction.

**Comment by aosewski** [comment]
@rosenrodt  regarding unit-tests structure. I see one more thing which could be done bit better - maybe next time I'll do that :) Part which is the most time consuming during unit-testing is input data initialization and host verification. We should be able to pass input tensors as well as reference results to profiler functions. Although within single profiler call those things are done once, but in multiple UT cases scenarios this is obvious waste of time. And we could leverage Fixtures to reuse those data.

**Comment by aosewski** [comment]
@rosenrodt  Could you please once again approve this PR? I just updated it with recent changes on develop.
@zjing14  Could you merge this PR?

**Review by rosenrodt — DISMISSED** [review]

File: library/src/tensor_operation_instance/gpu/softmax/device_softmax_f16_f16_instance_rank3_reduce1.cpp
Comment (qianfengz) [inline library/src/tensor_operation_instance/gpu/softmax/device_softmax_f16_f16_instance_rank3_reduce1.cpp:7]: Do we really need to include `device_softmax_f16_f16_instance_rankX_reduceY.hpp` by `device_softmax_f16_f16_instance_rankX_reduceY.cpp` ?

File: library/src/tensor_operation_instance/gpu/softmax/device_softmax_f16_f16_instance_rank3_reduce2.cpp
Comment (qianfengz) [inline library/src/tensor_operation_instance/gpu/softmax/device_softmax_f16_f16_instance_rank3_reduce2.cpp:17]: Can you share the naming from `library/include/ck/library/tensor_operation_instance/device_operation_instance_factory.hpp`

File: library/src/tensor_operation_instance/gpu/softmax/device_softmax_f16_f16_instance_rank3_reduce2.cpp
Comment (qianfengz) [inline library/src/tensor_operation_instance/gpu/softmax/device_softmax_f16_f16_instance_rank3_reduce2.cpp:18]: Add this line to `device_operation_instance_factory.hpp` if you don't want to use `PassThrough`

File: test/softmax/test_softmax_util.hpp
Comment (qianfengz) [inline test/softmax/test_softmax_util.hpp:17]: Many of the header files included here are not needed directly by this file

Comment (aosewski) [reply test/softmax/test_softmax_util.hpp]: Fixed by [2f6ab84](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/406/commits/2f6ab840abf8b7c3b138ea1cae6de6fa3f2b8d3f)

**Review by qianfengz — APPROVED** [review]

**Comment by zjing14** [comment]
@asroy This PR is ready for merge.

**Review by qianfengz — APPROVED** [review]

File: include/ck/tensor_operation/gpu/device/impl/device_softmax_impl.hpp
Comment (qianfengz) [inline include/ck/tensor_operation/gpu/device/impl/device_softmax_impl.hpp:287]: `reduce_lowest_length` is not defined, which causes compiling error.   Not completely consistent synchronization with the latest develop branch

**Review by qianfengz — CHANGES_REQUESTED** [review]

File: include/ck/tensor_operation/gpu/device/impl/device_softmax_impl.hpp
Comment (qianfengz) [reply include/ck/tensor_operation/gpu/device/impl/device_softmax_impl.hpp:287]: Keep use `p_arg_->inLengths_[Rank - 1]`

Comment (aosewski) [reply include/ck/tensor_operation/gpu/device/impl/device_softmax_impl.hpp]: Oh, I resolved merge conflicts too fast :) Fixed [here](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/406/commits/e39b065b550c34ca98cd3c927dffa69ff43aceab)

**Comment by qianfengz** [comment]
@asroy  This P.R can be merged

**Review by qianfengz — CHANGES_REQUESTED** [review]
There is probably some logical issue in the `DeviceSoftmaxImpl::IsSupported()`, which causes test failing

File: include/ck/tensor_operation/gpu/device/impl/device_softmax_impl.hpp
Comment (qianfengz) [inline include/ck/tensor_operation/gpu/device/impl/device_softmax_impl.hpp:302]: These lines are not needed.   Which should use `if(arg.inLengths_[Rank - 1] % InDstVectorSize != 0)` and that is already included by lines above.   Sorry for my not very complete comments previously

Comment (qianfengz) [reply include/ck/tensor_operation/gpu/device/impl/device_softmax_impl.hpp:302]: You can check the codes in develop branch for correcting

**Comment by aosewski** [comment]
> There is probably some logical issue in the `DeviceSoftmaxImpl::IsSupported()`, which causes test failing

Actually it's not the issue of `IsSupportedArgument` function which itself is correct. [Those test](https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/7797ded18cd203f96200e5163ab7e485a0e643ae/test/softmax/test_softmax_interface.cpp#L18-L60) were failing because they expect that `Argument` will throw when trying to construct object with invalid arguments. Previously this was checked by `DeviceReduceMultiblock::Argument` class (see [comment](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/406/files#r992215922)) but after recent merge with develop this was overlooked.

**Comment by asroy** [comment]
@qianfengz @rosenrodt please review/approve and I will merge it

**Review by qianfengz — APPROVED** [review]

**Review by rosenrodt — APPROVED** [review]

**Comment by aosewski** [comment]
@asroy The PR is ready to merge

**Comment by qianfengz** [comment]
> @asroy The PR is ready to merge

@asroy This is ready to merge

**Pull request merged**
